### PR TITLE
tests: fix flaky integration test

### PR DESCRIPTION
**Pull request by marcelo-gonzalez**
the download_file() test might fail if download_file() runs before
the HTTP server gets a chance to start. Fix it by waiting until the server
is already listening to call download_file(). This requires moving the
tokio runtime block_on() call to a wrapper function so we don't have problems
starting a runtime inside a runtime in the test code

### CODE REVIEW

**Comment by marcelo-gonzalez** [comment]
flaky run is here: https://buildkite.com/nearprotocol/nearcore/builds/12012#_

I guess I don't really have any more evidence that this is what caused it, but I'm pretty sure it is, because if you add a one-second sleep right before starting the server in the current test, you get the same conn refused error

File: integration-tests/src/tests/test_download_file.rs [hunk 1]
```diff
@@ -5,33 +5,39 @@ use std::fs::File;
...
+    rx.await.unwrap();
+
     nearcore::config::download_file(
         &format!("http://localhost:{}", port),
         tmp_downloaded_file_path,
     )
+    .await
     .unwrap();
 
     let mut downloaded_file = File::open(tmp_downloaded_file_path).unwrap();
```
Comment (matklad) [inline integration-tests/src/tests/test_download_file.rs:43]: could use `std::fs::read` here

File: integration-tests/src/tests/test_download_file.rs [hunk 2]
```diff
@@ -5,33 +5,39 @@ use std::fs::File;
...
-        tokio::runtime::Runtime::new().unwrap().block_on(async move {
-            let addr = SocketAddr::from(([127, 0, 0, 1], port));
-            let make_svc =
-                make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(handle_request)) });
-            let server = Server::bind(&addr).serve(make_svc);
-            if let Err(e) = server.await {
-                eprintln!("server error: {}", e);
-            }
-        });
+    let (tx, rx) = tokio::sync::oneshot::channel();
```
Comment (matklad) [inline integration-tests/src/tests/test_download_file.rs:17]: Doesn't matter in this case, but in general the rule-of-thumb primitive to use for synchronization in tests is barrier: https://docs.rs/tokio/latest/tokio/sync/struct.Barrier.html. It more directly expresses "wait util the coordinating threads have finished setup work", and generalizes for more than one thread.

**Review by matklad — APPROVED** [review]
:heart: always great to fix flaky tests!

Also: does this need to be an `integraton-test`? It seems that this doesn't depend on any of the shared testing machinery, and can be a unit-test in the `nearcore` crate.

File: integration-tests/src/tests/test_download_file.rs (hunk 2 above)
Comment (marcelo-gonzalez) [reply integration-tests/src/tests/test_download_file.rs]: ah cool, that does seem more clear. might as well update it

**Comment by marcelo-gonzalez** [comment]
> Also: does this need to be an `integraton-test`? It seems that this doesn't depend on any of the shared testing machinery, and can be a unit-test in the `nearcore` crate.

hmm good point. Could maybe be good to do that in another PR

**Review by mm-near — APPROVED** [review]

**Pull request merged**